src/design/inputs/date.js

```diff
@@ -25,48 +26,35 @@ export const DateInput = ({
 }: Props) => {
   const date = getDateFromString(value)
   const [isValid, setValid] = useState(Boolean(date))
-  const [day, setDay] = useState<number | null>(null)
-  const [month, setMonth] = useState<number | null>(null)
+  const [day, setDay] = useState<string | null>(null)
+  const [month, setMonth] = useState<string | null>(null)
   const [year, setYear] = useState<number | null>(null)
-  useEffect(() => {
-    if (date) {
-      setDay(date.get('date'))
-      setMonth(date.get('month') + 1)
-      setYear(date.get('year'))
-    }
-  }, [value])
 
   const onDayChange = (val) => {
-    const newDay = Number(val)
-    if (newDay > 31) return
-    const newDate = getDateFromVals(year, month, newDay)
-    setDay(newDay)
-    if (newDate) {
-      onChange(getStringFromDate(newDate))
-      setValid(true)
-    } else {
-      setValid(false)
-      onChange('')
-    }
+    const newDay = Number(val);
+    if (isNaN(newDay) || newDay > 31 || val.length > 2) return;
+    const newDate = getDateFromVals(year, month, newDay);
+    setDay(val);
+    onDateChange(newDate);
   }
+
   const onMonthChange = (val) => {
     const newMonth = Number(val)
-    if (newMonth > 12) return
+    if (isNaN(newMonth) || newMonth > 12 || val.length > 2) return
     const newDate = getDateFromVals(year, newMonth, day)
-    setMonth(newMonth)
-    if (newDate) {
-      onChange(getStringFromDate(newDate))
-      setValid(true)
-    } else {
-      setValid(false)
-      onChange('')
-    }
+    setMonth(val)
+    onDateChange(newDate)
   }
+
   const onYearChange = (val) => {
     const newYear = Number(val)
     if (newYear > new Date().getFullYear() + 3) return
     const newDate = getDateFromVals(newYear, month, day)
     setYear(newYear)
+    onDateChange(newDate)
+  }
+
+  const onDateChange = (newDate) => {
```

Comment:
```suggestion
  const dateChange = (newDate) => {
```
Not a big deal but I would usually reserve function names starting with "on" for event handlers. Refactoring all of the duplicate code into a single method is a great change but since we don't, for example, pass that function to another component to call when some specific behaviour happens, we shouldn't use a name that implies that is what it is used for. It's more like a helper function local to this component.

Reply:
Fair point! Changed this.